File: app/src/main/java/tik/prometheus/mobile/ui/home/HomeFragment.java
package tik.prometheus.mobile.ui.home;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import tik.prometheus.mobile.R;
import tik.prometheus.mobile.databinding.FragmentHomeBinding;
import tik.prometheus.mobile.services.HTTPConnector;
import tik.prometheus.mobile.utils.Utils;
import tik.prometheus.mobile.workflow.WorkflowRepos;

import java.util.ArrayList;
import java.util.List;

public class HomeFragment extends Fragment {

    private HomeViewModel homeViewModel;
    private FragmentHomeBinding binding;
    private WorkflowRepos workflowRepos;
    LineChart mpLinechart;

    @Override
    public void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        homeViewModel = new ViewModelProvider(this).get(HomeViewModel.class);
        workflowRepos = HTTPConnector.createService(WorkflowRepos.class, "VYNKK", "1");
        binding = FragmentHomeBinding.inflate(inflater, container, false);

        initChart(binding);
        LinearLayout sensorDashboard = binding.sensorDashboard;
        for (int i = 0; i <= 10; i++) {
            CardView card = new CardView(getContext());

            LinearLayout.LayoutParams layout = new LinearLayout.LayoutParams(200, 200);
            layout.setMargins(5, 5, 5, 5);
            card.setLayoutParams(layout);

            int color = Utils.Companion.themeColor(getContext(), R.attr.colorError);
            card.setCardBackgroundColor(color);

            sensorDashboard.addView(card);
        }
        final TextView textView = binding.textHome;
        homeViewModel.getText().observe(getViewLifecycleOwner(), new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                textView.setText(s);
            }
        });
        return binding.getRoot();
    }

    List<Entry> dataValues() {
        List<Entry> dataVals = new ArrayList<>();
        dataVals.add(new Entry(0, 20));
        dataVals.add(new Entry(1, 10));
        dataVals.add(new Entry(3, 30));
        dataVals.add(new Entry(9, 20));
        dataVals.add(new Entry(7, 10));
        return dataVals;
    }

    private void initChart(FragmentHomeBinding binding) {
        mpLinechart = binding.lineChart;
        List<ILineDataSet> dataSets = new ArrayList<>();
        dataSets.add(
                new LineDataSet(dataValues(), "Data set 1")
        );
        LineData data = new LineData(dataSets);
        mpLinechart.setData(data);
        mpLinechart.invalidate();
        int color = Utils.Companion.themeColor(getContext(), R.attr.colorOnBackground);
        mpLinechart.getAxisLeft().setTextColor(color);
        mpLinechart.getAxisRight().setTextColor(color);
        mpLinechart.getLegend().setTextColor(color);
        mpLinechart.getLineData().setValueTextColor(color);
        mpLinechart.getDescription().setTextColor(color);
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/tik/prometheus/mobile/ui/screen/home/HomeFragment.kt
package tik.prometheus.mobile.ui.screen.home

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import tik.prometheus.mobile.NestableFragment
import tik.prometheus.mobile.R
import tik.prometheus.mobile.databinding.FragmentHomeBinding
import tik.prometheus.mobile.ui.adapters.SensorAdapter
import tik.prometheus.mobile.ui.adapters.ZLoadStateAdapter
import tik.prometheus.mobile.ui.screen.sensor.SensorDetailFragment
import tik.prometheus.mobile.utils.themeColor

class HomeFragment : Fragment(), NestableFragment<SensorModel.SensorItem> {
    val TAG = HomeFragment::class.toString()

    private lateinit var viewModel: HomeViewModel
    private var _binding: FragmentHomeBinding? = null
    private val binding get() = _binding!!
    private var mpLinechart: LineChart? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        viewModel = ViewModelProvider(this).get(HomeViewModel::class.java)
        _binding = FragmentHomeBinding.inflate(layoutInflater, container, false)
        // CHAR
        initChart()

        // SENSOR
        val sensorAdapter = SensorAdapter(this)

        binding.sensorsRecyclerView.apply {
            layoutManager = LinearLayoutManager(context)
            setHasFixedSize(true)
            adapter = sensorAdapter.withLoadStateFooter(
                footer = ZLoadStateAdapter { sensorAdapter.retry() }
            )
        }
        lifecycleScope.launch {
            viewModel.sensors.collectLatest {
                sensorAdapter.submitData(it)
            }
        }
        binding.btnRetry.setOnClickListener {
            sensorAdapter.retry()
        }
        sensorAdapter.addLoadStateListener { loadState ->
            if (loadState.refresh is LoadState.Loading) {
                binding.btnRetry.visibility = View.GONE
                binding.progressBar.visibility = View.VISIBLE
            } else {
                binding.progressBar.visibility = View.GONE
                val errorState = when {
                    loadState.append is LoadState.Error -> loadState.append as LoadState.Error
                    loadState.prepend is LoadState.Error -> loadState.prepend as LoadState.Error
                    loadState.refresh is LoadState.Error -> {
                        binding.btnRetry.visibility = View.VISIBLE
                        loadState.refresh as LoadState.Error
                    }

                    else -> null
                }
                errorState?.let {
                    showToast(it.error.message.toString())
                }
            }
        }

        return binding.root
    }
//
//
//    private fun initSensorDashboard() {
//        val sensorRecycleView = binding!!.sensorsRecyclerView
//        sensorRecycleView.layoutManager = LinearLayoutManager(this.context, LinearLayoutManager.VERTICAL, false)
//        val sensorAdapter = SensorAdapter {}
//        sensorRecycleView.adapter = sensorAdapter
//        val restServiceApi = RestServiceHelper.getInstance().create(RestServiceApi::class.java)
//        GlobalScope.launch {
//            val queryMap = Pageable(0, 10, ArrayList()).toMap()
//            val res = restServiceApi.getSensors(queryMap)
//            if (res.isSuccessful) {
//                val sensorPage = res.body();
//                for (sensor: Sensor in sensorPage?.content!!) {
//                    val card = CardView(requireContext())
//                    val layout = LinearLayout.LayoutParams(200, 200)
//                    layout.setMargins(5, 5, 5, 5)
//                    card.layoutParams = layout
//                    val color: Int = requireContext().themeColor(R.attr.colorError)
//                    card.setCardBackgroundColor(color)
//                    val txt = TextView(requireContext())
//                    txt.text = sensor.topic
//                    card.addView(txt)
//                    viewModel?.sensorMap?.value?.put(sensor.id, sensor)
//                    sensorRecycleView.addView(card)
//                }
//            }
//        }
//    }

    private fun initChart() {
        mpLinechart = binding.lineChart
        val dataSets: MutableList<ILineDataSet> = ArrayList()
        dataSets.add(
            LineDataSet(dataValues(), "Data set 1")
        )
        val data = LineData(dataSets)
        mpLinechart!!.data = data
        mpLinechart!!.invalidate()
        val color: Int = requireContext().themeColor(R.attr.colorOnBackground)
        mpLinechart!!.axisLeft.textColor = color
        mpLinechart!!.axisRight.textColor = color
        mpLinechart!!.legend.textColor = color
        mpLinechart!!.lineData.setValueTextColor(color)
        mpLinechart!!.description.textColor = color
    }

    fun dataValues(): List<Entry> {
        val dataVals: MutableList<Entry> = ArrayList()
        dataVals.add(Entry(0f, 20f))
        dataVals.add(Entry(1f, 10f))
        dataVals.add(Entry(3f, 30f))
        dataVals.add(Entry(9f, 20f))
        dataVals.add(Entry(7f, 10f))
        return dataVals
    }


    private fun showToast(value: CharSequence) {
        Toast.makeText(this.context, value, Toast.LENGTH_LONG).show()
    }

    override fun insertNestedFragment(model: SensorModel.SensorItem) {
        val childFragment = SensorDetailFragment(model);
        activity?.supportFragmentManager?.let {
            it.beginTransaction()
                .replace(R.id.nav_host_container, childFragment)
                .addToBackStack(SensorDetailFragment::class.java.name)
                .commit()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

}